### ISSUE TITLE
Allow manual configuration of the native plugins for Flutter

### ISSUE THREAD

**Issue by kuhnroyal** [getsentry/sentry-dart#265]
Currently errors that occur prior to the initialization of the Flutter engine or in background services that have no Flutter engine started, can not be send to Sentry because the native plugins are not initialized yet.

These are rare cases but if they are required for a particular app, then it would be great to have the additional option of manual configuration of the native plugins.

See https://github.com/getsentry/sentry-dart/discussions/255 for more discussions on this topic.

**Comment by bruno-garcia** [comment]
Recently was added to RN a way to opt-out of initializing the native layer through the JS layer. 
I believe a solution here could be rather simple: We have a escape hatch to Dart initializing the native layer, as we did in RN. And clearly document and support a way that users can be "on their own" with regards to initializing things. `init` would need to be called manually in all layers.

RN issue; https://github.com/getsentry/sentry-react-native/issues/542

**Comment by marandaneto** [comment]
there's a gotcha on Android, the Flutter engine is initialized via ContentProvider which is run before the Application class, so even if initializing the Android SDK manually, you won't be able to capture those errors, but only after SDK is fully initialized.

If you decide to init. the Android SDK via ContentPovider too (our auto init-mode works this way), the ContentProviders initialize in a sort of random order, so likely won't be capturing errors either, it's tricky, not sure if there's an easy way out of it.

it might be possible to somehow initialize them in order if using the `startup` package, https://developer.android.com/reference/androidx/startup/package-summary
but it'd require code on the client's side, didn't test though.

**Comment by btrautmann** [comment]
Hi there!

We're doing some investigation into relying solely on the `sentry_flutter` SDK for crash/issue reporting in our _embedded_ application, and I figured here would be a good place to clarify some of our questions/observations:

### Questions:

1. When using `sentry_flutter`, is it expected to include the native SDK as well (i.e use both `sentry_flutter` and `sentry-android`, even if we're not calling `SentryAndroid.init`)? Is there any documentation about the expected setup?
2. I assume that in light of https://github.com/getsentry/sentry-dart/pull/765, the current expectation is to NOT call `SentryAndroid.init` if you're also calling `SentryFlutter.init` on the Flutter side (I'm asking primarily in the add-to-app use case). Is that correct?
3. We were previously using `sentry-android-timber` to integrate with `Timber`, but have successfully dropped that in favor of sending manual `Event`s and `Breadcrumb`s over a `MethodChannel` to be reported by Flutter. Is this an expected use case? My understanding is that _only_ **automatic** breadcrumbs and crashes will be caught by Flutter (setting aside any of the initialization edge cases mentioned above), and that anything **manual** would need to be communicated to Flutter to then be reported.

### Observations:

1. I'm noticing that when doing the things mentioned in Question 2 above (sending `Event`s and `Breadcrumb`s over a `MethodChannel`, the `Breadcrumb`s only show up in the Sentry UI when an `Event` is triggered, but NOT when an uncaught `Exception` occurs, i.e:
```kotlin
// Goes to a custom SentryTree that sends a breadcrumb payload to Flutter which reports to Sentry
Timber.w("my breadcrumb that will be communicated over a MethodChannel")
...
throw Exception("Oh no!")
```
☝️  In the above case, the `Breadcrumb` created by the call to `Timber.w` will NOT display in the `Breadcrumb`s for that `Exception`, _but_ if I do the following
```kotlin
// Goes to a custom SentryTree that sends a breadcrumb payload to Flutter which reports to Sentry
Timber.w("my breadcrumb that will be communicated over a MethodChannel")
...
// Goes to the custom SentryTree that sends an event payload to Flutter which reports to Sentry
Timber.e("An event occured!")
```
☝️  In the above case, the `Event` created by the call to `Timber.e` WILL display in the `Breadcrumb`s for that `Event`.

Is this expected, or is that a bug?

I think what I'm most looking for is actual documentation regarding what the "typical" setup of an add-to-app Flutter application is, and maybe how this is expected to change in light of https://github.com/getsentry/sentry-dart/pull/765.

Thank you!

**PS: If this would be better created as a separate issue, please let me know and I can do so!**

**Comment by marandaneto** [comment]
> Hi there!
> 
> We're doing some investigation into relying solely on the `sentry_flutter` SDK for crash/issue reporting in our _embedded_ application, and I figured here would be a good place to clarify some of our questions/observations:
> 
> ### Questions:
> 1. When using `sentry_flutter`, is it expected to include the native SDK as well (i.e use both `sentry_flutter` and `sentry-android`, even if we're not calling `SentryAndroid.init`)? Is there any documentation about the expected setup?

Yes, https://docs.sentry.io/platforms/flutter/#configure
All you need to do is to call `SentryFlutter.init`, we do everything else automatically.

> 2. I assume that in light of [Feat: Allow manual init of the Native SDK or no Native SDK at all #765](https://github.com/getsentry/sentry-dart/pull/765), the current expectation is to NOT call `SentryAndroid.init` if you're also calling `SentryFlutter.init` on the Flutter side (I'm asking primarily in the add-to-app use case). Is that correct?

Correct, you don't need to call `SentryAndroid. init`, after #765 gets merged and released, you may opt-out of this behavior in case you want custom configuration on the native side (Android/iOS).

> 3. We were previously using `sentry-android-timber` to integrate with `Timber`, but have successfully dropped that in favor of sending manual `Event`s and `Breadcrumb`s over a `MethodChannel` to be reported by Flutter. Is this an expected use case? My understanding is that _only_ **automatic** breadcrumbs and crashes will be caught by Flutter (setting aside any of the initialization edge cases mentioned above), and that anything **manual** would need to be communicated to Flutter to then be reported.

Yes, if you want e.g. breadcrumbs on the Flutter side, just do `Sentry.addBreadcrumb(...)`.
If you want a breadcrumb on Native side, e.g. Android, you have to set up your own MethodChannel and call the Native SDK there, but the API is also called `Sentry.addBreadcrumb(...)`.

> ### Observations:
> 1. I'm noticing that when doing the things mentioned in Question 2 above (sending `Event`s and `Breadcrumb`s over a `MethodChannel`, the `Breadcrumb`s only show up in the Sentry UI when an `Event` is triggered, but NOT when an uncaught `Exception` occurs, i.e:
> 
> ```kotlin
> // Goes to a custom SentryTree that sends a breadcrumb payload to Flutter which reports to Sentry
> Timber.w("my breadcrumb that will be communicated over a MethodChannel")
> ...
> throw Exception("Oh no!")
> ```
> 
> ☝️ In the above case, the `Breadcrumb` created by the call to `Timber.w` will NOT display in the `Breadcrumb`s for that `Exception`, _but_ if I do the following
> 
> ```kotlin
> // Goes to a custom SentryTree that sends a breadcrumb payload to Flutter which reports to Sentry
> Timber.w("my breadcrumb that will be communicated over a MethodChannel")
> ...
> // Goes to the custom SentryTree that sends an event payload to Flutter which reports to Sentry
> Timber.e("An event occured!")
> ```
> 
> ☝️ In the above case, the `Event` created by the call to `Timber.e` WILL display in the `Breadcrumb`s for that `Event`.
> 
> Is this expected, or is that a bug?
> 
> I think what I'm most looking for is actual documentation regarding what the "typical" setup of an add-to-app Flutter application is, and maybe how this is expected to change in light of #765.
> 
> Thank you!
> 
> **PS: If this would be better created as a separate issue, please let me know and I can do so!**

To understand that better, I'd need to know what your custom `SentryTree` does, but I have a hunch.
When you do `throw Exception("Oh no!")`, Flutter actually swallow your exception and the event is sent by the Flutter layer and not the Native layer, but your breadcrumb was added only on the Native layer.

When you do `Timber.e("An event occured!")`, you are actually calling `Sentry.captureException` most likely on the Native layer, hence the breadcrumb lives there.

Am I right to assume that? This would ideally be fixed by https://github.com/getsentry/sentry-dart/issues/194 which is a sort of extension for #765 as well, which means, when you call `Sentry.addBreadcrumb(...)` on Flutter, this breadcrumb is automatically synced to the Native layer via MethodChannel, automatically.

**Comment by btrautmann** [comment]
> To understand that better, I'd need to know what your custom `SentryTree` does, but I have a hunch. When you do `throw Exception("Oh no!")`, Flutter actually swallow your exception and the event is sent by the Flutter layer and not the Native layer, but your breadcrumb was added only on the Native layer.
> 
> When you do `Timber.e("An event occured!")`, you are actually calling `Sentry.captureException` most likely on the Native layer, hence the breadcrumb lives there.
> 
> Am I right to assume that? This would ideally be fixed by #194 which is a sort of extension for #765 as well, which means, when you call `Sentry.addBreadcrumb(...)` on Flutter, this breadcrumb is automatically synced to the Native layer via MethodChannel, automatically.

Sorry for not being more clear -- I've put together a diagram of our intention. Currently we're running an add-to-app setup but are moving towards a purely Flutter application. As such, we're trying to lift as many components over to Flutter as possible so we can rid ourselves of our legacy native codebase. With that in mind, here's what we're picturing (let me know if anything in this diagram looks off or incorrect):

![Untitled-2022-02-22-1707](https://user-images.githubusercontent.com/8343465/156586554-5f18ee6c-a1c0-46a6-9a62-e903be00bbaa.png)

So basically, the hope would be:

1. No sentry dependencies on the native side
2. Flutter depends on `sentry_flutter` and sets things up via `SentryFlutter.init`
3. Native hooks into info, warning, and error logs via `Timber` and just passes the metadata to Flutter via a `MethodChannel`
4. Flutter invokes `addBreadcrumb()` and `captureEvent()` per our business logic when it receives the metadata from the `MethodChannel`
5. When an `Event` or uncaught `Exception` occurs, we see these things in our Sentry Flutter project (i.e we see all `Breadcrumb`s that were sent over the `MethodChannel` from Native _and/or_ the ones added directly from Flutter.

However, like I described above, we're not seeing the `Breadcrumbs` that were reported to Flutter when an `Exception` occurs on the native side, but we DO see them if I simply report and `Event`.
